### ISSUE TITLE
Color artifacts with fixed color shader

### ISSUE THREAD

**Issue by samraul** [facebookresearch/pytorch3d#1733]
## 🐛 Bugs / Unexpected behaviors

I have color artifacts rendering with a fixed color shader.

* I have a custom `UnlitColorShader` to render with `hard_rgb_blend` which I have successfully used previously (no artifacts).
* I am using both Orthographic and Perspective cameras, and both have the same issue.
* I have been able to reproduce it with simple planes.
* I have tried `blur_radius=0.00001`, `bin_size=0` and `faces_per_pixel=1`, as responses in similar issues suggested, but the problem persists.
* I have looked for similar issues such as https://github.com/facebookresearch/pytorch3d/issues/1393, https://github.com/facebookresearch/pytorch3d/issues/1026, https://github.com/facebookresearch/pytorch3d/issues/516, https://github.com/facebookresearch/pytorch3d/issues/471, but none of the solutions have worked so far.

| As color  | As mask (same as color) |
| ------------- | ------------- |
| <kbd><img width="600" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/da92c34f-f4ff-4b04-a257-548423d33bb9"/></kbd> | <kbd><img width="600" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/0c259965-5fb1-4f9e-8a5c-3cb167ab0c2d"/></kbd> |
(Images are with perspective camera. Code below with orthographic camera has the same issue).

## Instructions To Reproduce the Issue:

1. Any changes you made (`git diff`) or code you wrote
```python
class UnlitColorShader(ShaderBase):
    """Shader that paints each face with the colors of its vertices, with no lighting."""

    def __init__(self, device="cpu", blend_params=None):
        super().__init__(device=device, blend_params=blend_params)

    def forward(self, fragments, meshes, **kwargs) -> torch.Tensor:
        blend_params = kwargs.get("blend_params", self.blend_params)
        texels = meshes.sample_textures(fragments)
        images = hard_rgb_blend(texels, fragments, blend_params)
        return images  # (N, H, W, 3) 
```
2. The exact command(s) you ran:

```python

  # -- Camera config
  R, T = look_at_view_transform(
      eye=((camera_position.x, camera_position.y, camera_position.z),),
      at=((camera_target.x, camera_target.y, camera_target.z),),
  )
  SCENE_SCALE = 1
  xy_extent = 0.5
  cameras = FoVOrthographicCameras(
      device=renderer.device,
      R=R,
      T=T,
      znear=0.001 * SCENE_SCALE,
      zfar=camera_pos.z * 1.1 * SCENE_SCALE,
      min_x=-SCENE_SCALE * xy_extent,
      max_x=SCENE_SCALE * xy_extent,
      min_y=-SCENE_SCALE * xy_extent,
      max_y=SCENE_SCALE * xy_extent,
  )

  ...
  # -- Relevant color config
  # Each quad has an assigned color, shared by the 4 vertices of the quad
  vertex_colors = np.array([fixed_colors[quad_type] * 4 for quad_type in mesh.quad_types])
  vertex_colors = vertex_colors.reshape(-1, 3)
  verts_features = torch.tensor(vertex_colors, dtype=torch.float32).unsqueeze(0)
  textures = TexturesVertex(verts_features=verts_features)  # `verts_features` expected to be (N, V, 3)

  ...
  # -- Rasterization code
  blend_params = BlendParams(background_color=[0,0,0])
  color_shader = UnlitColorShader(device=self.device, blend_params=blend_params)
  raster_settings = RasterizationSettings(image_size=self.image_size, blur_radius=0.0, faces_per_pixel=1)
  rasterizer = MeshRasterizer(cameras=cameras, raster_settings=raster_settings)
  renderer = MeshRenderer(rasterizer=rasterizer, shader=color_shader)
  images = renderer(self.meshes.to(self.device))
```

3. What you observed (including the full logs):

I have successfully used similar code with similar parameters before, so I am unsure what can cause the given artifacts.
I am using `pytorch3d==0.7.3`.

Any hints would be appreciated.

Thank you!

**Comment by bottler** [comment]
I don't know what you are expecting. What is the artifact?

Can you paste an image of the mesh, e.g. with plotly_vis, so we can see what it is?

**Comment by samraul** [comment]
Thanks, @bottler.

Certainly, let me clarify:

* Left image: generated image with supposedly 3 hard rgb colors
* Center image: mask computed from the left image, which shows the artifacts (some pixels have minor deviations)
* Right image: expected mask 

| Generated image | Computed mask | Expected mask |
| ------------- | ------------- | ------------- |
| <kbd><img width="300" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/da92c34f-f4ff-4b04-a257-548423d33bb9"/></kbd> | <kbd><img width="300" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/0c259965-5fb1-4f9e-8a5c-3cb167ab0c2d"/></kbd> | <kbd><img width="300" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/1511c769-53fe-48bc-bb19-08c517a362a9"/></kbd> |

```
    mask = np.any(
        [np.all(generated_image == color[category], axis=-1) for category in [C1, C2, C2]],
        axis=0,
    )
```

---

Both these pixels should be `[80, 40, 60]`, however the one on the right shows `[79, 39, 59]`
| Good color  | Bad color |
| ------------- | ------------- |
| <kbd><img width="400" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/893c3d31-388c-4cf9-80ed-153fe269387c"/></kbd> | <kbd><img width="400" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/ec490bef-4fd4-45d7-837e-7efc1cad86b6"/></kbd> |

---

Here is the mesh, lit, and from a different angle. At the moment it is simply a few quads, each of which has a unique color:

<kbd><img width="600" src="https://github.com/facebookresearch/pytorch3d/assets/73083942/4e03dcae-c55b-4878-be57-734e80ed9dbc"/></kbd>

It looks like a precision issue, perhaps converting from normalized color to uint8. But I have never observed this before despite having used it in different meshes.

**Comment by samraul** [comment]
Ah, it definitely looks like a precision issue.
I removed all planes except one, expecting only 4 unique values, however `print(torch.unique(images))` shows:
```
tensor([0.0000, 0.1569, 0.1569, 0.1569, 0.1569, 0.1569, 0.1569, 0.2353, 0.2353,
        0.2353, 0.2353, 0.2353, 0.2353, 0.2353, 0.3137, 0.3137, 0.3137, 0.3137,
        0.3137, 0.3137, 1.0000], device='cuda:0')
```

**Comment by samraul** [comment]
Adding a small epsilon before conversion solves the issue.
````python
        print("-" * 20)
        print(torch.unique(images))
        print(torch.unique(images * 255))

        print("-" * 20)
        epsilon = 0.0001
        as_np = images.cpu().squeeze().numpy()
        as_np_rounded_8 = np.round(as_np * 255 + epsilon).astype(np.uint8)        
        print(np.unique(as_np_rounded_8))
```

```
--------------------
tensor([0.0000, 0.1569, 0.1569, 0.1569, 0.1569, 0.1569, 0.1569, 0.2353, 0.2353,
        0.2353, 0.2353, 0.2353, 0.2353, 0.2353, 0.3137, 0.3137, 0.3137, 0.3137,
        0.3137, 0.3137, 1.0000], device='cuda:0')
tensor([  0.0000,  40.0000,  40.0000,  40.0000,  40.0000,  40.0000,  40.0000,
         60.0000,  60.0000,  60.0000,  60.0000,  60.0000,  60.0000,  60.0000,
         80.0000,  80.0000,  80.0000,  80.0000,  80.0000,  80.0000, 255.0000],
       device='cuda:0')
--------------------
[  0  40  60  80 255]
```

If anyone has a better way, please let me know, otherwise we can close the issue, since this is just inherent to python type conversions.

**Comment by samraul** [comment]
👍  Will close and leave here in case someone runs into similar artifacts. Thanks.
